src/flexible_credential_source.rs

```diff
@@ -0,0 +1,297 @@
+        }
+    }
+
+    pub(crate) async fn from_default_credentials() -> Result<Self, Error> {
+        tracing::debug!("Loading user credentials file");
+        let mut home = dirs_next::home_dir().ok_or(Error::NoHomeDir)?;
+        home.push(Self::USER_CREDENTIALS_PATH);
+        Self::from_file(home).await
+    }
+
+    pub(crate) async fn try_into_service_account(
+        self,
+        client: &HyperClient,
+    ) -> Result<Box<dyn ServiceAccount>, Error> {
+        match self {
+            FlexibleCredentialSource::ServiceAccount(creds) => {
+                let service_account = CustomServiceAccount::new(creds)?;
+                Ok(Box::new(service_account))
+            }
+            FlexibleCredentialSource::AuthorizedUser(creds) => {
+                let service_account =
+                    ConfigDefaultCredentials::from_user_credentials(creds, client).await?;
+                Ok(Box::new(service_account))
+            }
+            FlexibleCredentialSource::ImpersonatedServiceAccount(creds) => {
+                let source_creds: Box<dyn ServiceAccount> = match *creds.source_credentials {
+                    FlexibleCredentialSource::AuthorizedUser(creds) => {
+                        let service_account =
+                            ConfigDefaultCredentials::from_user_credentials(creds, client).await?;
+                        Box::new(service_account)
+                    }
+                    FlexibleCredentialSource::ServiceAccount(creds) => {
+                        let service_account = CustomServiceAccount::new(creds)?;
+                        Box::new(service_account)
+                    }
+                    FlexibleCredentialSource::ImpersonatedServiceAccount(_) => {
+                        return Err(Error::NestedImpersonation)
+                    }
+                };
+
+                let service_account = ImpersonatedServiceAccount::new(
+                    source_creds,
+                    creds.service_account_impersonation_url,
+                    creds.delegates,
+                );
+
+                Ok(Box::new(service_account))
+            }
+        }
+    }
+
+    /// Read service account credentials from the given JSON file
+    async fn from_file<T: AsRef<Path>>(path: T) -> Result<Self, Error> {
+        let creds_string = fs::read_to_string(&path)
+            .await
+            .map_err(Error::UserProfilePath)?;
+
+        serde_json::from_str::<FlexibleCredentialSource>(&creds_string)
+            .map_err(Error::CustomServiceAccountCredentials)
+    }
+}
+
+// This credential uses the `source_credentials` to get a token
+// and then uses that token to get a token impersonating the service
+// account specified by `service_account_impersonation_url`.
+// refresh logic https://github.com/golang/oauth2/blob/a835fc4358f6852f50c4c5c33fddcd1adade5b0a/google/internal/externalaccount/impersonate.go#L57
+#[derive(Deserialize, Debug)]
+pub(crate) struct ImpersonatedServiceAccountCredentials {
```

Comment:
The only weirdness about this PR is that we put the deserialization for `ImpersonatedServiceAccount` inside the `flexible_credential_source` file. I did this because in theory the source credentials could be anything (the referenced go sdk source didn't make any descrimination).

If there is a cleaner approach to this, please let me know.